### PR TITLE
migrate from Narrow to SplitTable

### PR DESCRIPTION
**Pull request by hughperkins**
Using NVIDIA 940M and 5200U processor, this improves iteration times as follows:
- CUDA, cutorch: 0.13s => 0.11s
- OpenCL, cltorch: 0.27s => 0.26s
- CPU: 3.5s => 3.0s

### CODE REVIEW

**Comment by karpathy** [comment]
I assume you have also verified that it gives identical results as well? Haven't used these before

**Comment by hughperkins** [comment]
Training loss decreases.  After one epoch using cutorch, training loss is 2.09:

```
421/21150 (epoch 0.995), train_loss = 2.06398542, grad/param norm = 3.1865e-01, time/batch = 0.12s	
422/21150 (epoch 0.998), train_loss = 2.07728148, grad/param norm = 2.9893e-01, time/batch = 0.11s	
423/21150 (epoch 1.000), train_loss = 2.08978661, grad/param norm = 2.4922e-01, time/batch = 0.11s	
424/21150 (epoch 1.002), train_loss = 2.15074668, grad/param norm = 2.9246e-01, time/batch = 0.11s	
425/21150 (epoch 1.005), train_loss = 2.09268132, grad/param norm = 3.5920e-01, time/batch = 0.11s	
426/21150 (epoch 1.007), train_loss = 2.05545285, grad/param norm = 4.0142e-01, time/batch = 0.11s	
```
After 1000 iterations, 1.72:
```
1017/21150 (epoch 2.404), train_loss = 1.72333642, grad/param norm = 1.9213e-01, time/batch = 0.11s	
1018/21150 (epoch 2.407), train_loss = 1.73758396, grad/param norm = 1.6325e-01, time/batch = 0.11s	
1019/21150 (epoch 2.409), train_loss = 1.70474525, grad/param norm = 1.4151e-01, time/batch = 0.11s	
1020/21150 (epoch 2.411), train_loss = 1.65908576, grad/param norm = 1.2490e-01, time/batch = 0.11s	
1021/21150 (epoch 2.414), train_loss = 1.73056459, grad/param norm = 1.2686e-01, time/batch = 0.11s	
1022/21150 (epoch 2.416), train_loss = 1.73195712, grad/param norm = 1.3547e-01, time/batch = 0.11s	
1023/21150 (epoch 2.418), train_loss = 1.70630633, grad/param norm = 1.5645e-01, time/batch = 0.11s	
1024/21150 (epoch 2.421), train_loss = 1.67318447, grad/param norm = 1.7056e-01, time/batch = 0.11s	
1025/21150 (epoch 2.423), train_loss = 1.62619691, grad/param norm = 1.5619e-01, time/batch = 0.11s	
1026/21150 (epoch 2.426), train_loss = 1.66633054, grad/param norm = 2.0736e-01, time/batch = 0.11s	
1027/21150 (epoch 2.428), train_loss = 1.68936435, grad/param norm = 2.1618e-01,
```
Sample at that point:
```
user@pear:~/git/char-rnn-clean$ ls -lh cv/
total 2.0M
-rw-rw-r-- 1 user user 2.0M Jul 25 21:35 lm_lstm_epoch2.36_1.7218.t7
user@pear:~/git/char-rnn-clean$ samba-^C
user@pear:~/git/char-rnn-clean$ th sample.lua cv/lm_lstm_epoch2.36_1.7218.t7 
nn.Narrow	
using CUDA on GPU 0...	
creating an LSTM...	
missing seed text, using uniform probability over first character	
--------------------------	
gon:
This istering in the boonfeee.

PRISHIO:
Reseep leary, Ore man lanothing me: and love andred,
Therefore Godathoul, shear thou his canmbant ot,-
He pome to here, herefore ene, soy
Kis o, the secith of deserves one him;
Ulfoch not me
putine.

LEUCTIO:
Oo love any hroke Catitil,
Groom marior is that thisse!

PRUSUS:
Mather uscond, I come anjestry a mastans;
And thetelare with him orlaous speam will rexore.

GRUMUO:
For hour much proforay, as what heart a candatate
Leave-min anesin.

JUULET:
That speat our proate of ow, mores of forth!
Tay, mole, ding to Holk? Are your gruentinglady?
Thou to I saan in the morress shages!
Whi, will Allowing siscery lock, lorging where is vistient,
Bone know-ploofe the secbing and speak and piritiest.

COMINHARD:
Than, you she dillian Backent me sir,
If how chatce, I mound of Herivinine; and here.

TresoNttry!
Not shall seak?

TAFRYO:
Me, what as will inters, is wan He couse,
If to peapon that?

POMPEYt:
You hould an tode ciphs evener you
Ha swenms list.
```

**Comment by karpathy** [comment]
that's good, but if you work with nns for a while you realize that they often work even in presence of bugs. Can you confirm that you get exactly the same numbers in this output with the older version? Since the seeds for torch/cutorch are set manually you should see exactly the same output. Sorry to be a pain about this, but things like this have bitten me in the past.

**Comment by hughperkins** [comment]
Ah, good idea.  And I agree that it's important to check, especially since there's a thread right now on reddit linking to char-rnn :-)  https://www.reddit.com/r/MachineLearning/comments/3ejizl/karpathy_charrnn_doubt/

**Comment by hughperkins** [comment]
Ok, so:

using master, with seed 123, on cutorch (can see it is master because time is 0.13s per iteration):
```
1000/21150 (epoch 2.364), train_loss = 1.69123832, grad/param norm = 1.8980e-01, time/batch = 0.13s	
1001/21150 (epoch 2.366), train_loss = 1.82341127, grad/param norm = 1.7749e-01, time/batch = 0.14s	
1002/21150 (epoch 2.369), train_loss = 1.69782269, grad/param norm = 1.5184e-01, time/batch = 0.13s	
1003/21150 (epoch 2.371), train_loss = 1.69308439, grad/param norm = 1.6510e-01, time/batch = 0.13s	
```

using `splittable` branch, seed 123, cutorch:
```
1000/21150 (epoch 2.364), train_loss = 1.69123832, grad/param norm = 1.8980e-01, time/batch = 0.11s	
1001/21150 (epoch 2.366), train_loss = 1.82341127, grad/param norm = 1.7749e-01, time/batch = 0.12s	
1002/21150 (epoch 2.369), train_loss = 1.69782269, grad/param norm = 1.5184e-01, time/batch = 0.11s	
1003/21150 (epoch 2.371), train_loss = 1.69308439, grad/param norm = 1.6510e-01, time/batch = 0.11s	
```

**Comment by karpathy** [comment]
Awesome, thanks!

**Comment by hughperkins** [comment]
Cool :-)
